##### Missing value Data

In [107]:
#adding missing values in dataframe : np.nan
import numpy as np
import pandas as pd
df=pd.DataFrame({"A":[1,2,np.nan],"B":[5,np.nan,np.nan],"c":[1,2,3]})  #np.nan is used to add missing values 
print(df)
#print(df[0:1])
print(df.loc[0])


     A    B  c
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  3
A    1.0
B    5.0
c    1.0
Name: 0, dtype: float64


In [4]:
df.isnull().sum()  #to compute total missing values in all the columns

A    1
B    2
c    0
dtype: int64

In [25]:
#creating dataFrame with some missing values  under age and height column
df1 = pd.DataFrame({"Age":[12,np.nan,14,np.nan,16,12,13,14,12,13],"Height":[130,np.nan,133,143,123,np.nan,np.nan,np.nan,120,121]})
df1

Age  Height
0  12.0   130.0
1   NaN     NaN
2  14.0   133.0
3   NaN   143.0
4  16.0   123.0
5  12.0     NaN
6  13.0     NaN
7  14.0     NaN
8  12.0   120.0
9  13.0   121.0

# handling missing values: delete or fill with some values


##### we can handle missing values  by 
1.filling with some values using fillna() method


2.dropping the value

##### for intergal data we can fill missing value either by mean or median 
we use median : when we have few outliers(low and high values) and mostly we have values in specific range otherwise use meean




###mostly median is used

#### for categorical data : we use mode (most frequent value)

In [7]:
df1["Age"].mean()


13.25

In [29]:
#editing df1 by adding gender column with some missing values
df1 = pd.DataFrame({"Age":[12,np.nan,14,np.nan,26,12,13,24,12,13],"Height":[130,np.nan,133,143,123,np.nan,np.nan,np.nan,120,121],"Gender":["m","f",np.nan,"m",np.nan,"f","m",np.nan,np.nan,"m"]})
df1


Age  Height Gender
0  12.0   130.0      m
1   NaN     NaN      f
2  14.0   133.0    NaN
3   NaN   143.0      m
4  26.0   123.0    NaN
5  12.0     NaN      f
6  13.0     NaN      m
7  24.0     NaN    NaN
8  12.0   120.0    NaN
9  13.0   121.0      m

In [12]:
df1["Age"].median()


13.0

###### filling missing values of "age" column with median 

In [30]:
df1.Age.fillna(df1.Age.median(),inplace=True)
df1

Age  Height Gender
0  12.0   130.0      m
1  13.0     NaN      f
2  14.0   133.0    NaN
3  13.0   143.0      m
4  26.0   123.0    NaN
5  12.0     NaN      f
6  13.0     NaN      m
7  24.0     NaN    NaN
8  12.0   120.0    NaN
9  13.0   121.0      m

In [33]:
df1.dtypes   # checking types of all columns

Age       float64
Height    float64
Gender     object
dtype: object

##### filling missing values of "gender"  column with its mode as it has categorical data 

In [35]:
df1["Gender"].mode()  #gives most frequent value(,ultiple modes can be there thats y it give series where indexing is there thts y , 0 is there

0    m
dtype: object

In [37]:
#mode1=df1["Gender"].mode()[0]   : it fectches m from index 
df1.Gender.fillna(df1["Gender"].mode()[0],inplace=True)
df1


Age  Height Gender
0  12.0   130.0      m
1  13.0     NaN      f
2  14.0   133.0      m
3  13.0   143.0      m
4  26.0   123.0      m
5  12.0     NaN      f
6  13.0     NaN      m
7  24.0     NaN      m
8  12.0   120.0      m
9  13.0   121.0      m

##### removing missing values :  dropna()

In [94]:
# importing the datset and checking no of rows and columns import pandas as pd
import numpy as np
data=pd.read_csv("D:\\python_datascience\\data sets\\upcoming-event.csv")
data.head()
print("no of rows and columns is ",data.shape)




no of rows and columns is  (9, 119)


In [95]:
# this data has 119 columns : checking null values count in each column
data.isnull().sum()

R_fighter     0
B_fighter     0
R_odds        0
B_odds        0
R_ev          0
             ..
b_dec_odds    1
r_sub_odds    0
b_sub_odds    1
r_ko_odds     0
b_ko_odds     1
Length: 119, dtype: int64

In [96]:
# here the data is big so we check null values for column 1 to 20
data.isnull().sum()[1:21]

B_fighter                0
R_odds                   0
B_odds                   0
R_ev                     0
B_ev                     0
date                     0
location                 0
country                  0
Winner                   9
title_bout               0
weight_class             0
gender                   0
no_of_rounds             0
B_current_lose_streak    0
B_current_win_streak     0
B_draw                   0
B_avg_SIG_STR_landed     0
B_avg_SIG_STR_pct        0
B_avg_SUB_ATT            0
B_avg_TD_landed          0
dtype: int64

In [97]:
#3 from above data we can see the column name winner has 9 null values meand in all rows  as follows so need to remove 
data.Winner

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
5   NaN
6   NaN
7   NaN
8   NaN
Name: Winner, dtype: float64

In [98]:

data.drop(columns=["Winner"],inplace=True)
data.shape  


(9, 118)

####  from above mentioned code  we found the column named winner which has 9 nan values so we deleted and check its shape is is showing 118 columns means winner column has been deleted

In [99]:
## fetching all columns name 
print(data.columns)
data.columns[1:20]


Index(['R_fighter', 'B_fighter', 'R_odds', 'B_odds', 'R_ev', 'B_ev', 'date',
       'location', 'country', 'title_bout',
       ...
       'finish_details', 'finish_round', 'finish_round_time',
       'total_fight_time_secs', 'r_dec_odds', 'b_dec_odds', 'r_sub_odds',
       'b_sub_odds', 'r_ko_odds', 'b_ko_odds'],
      dtype='object', length=118)


Index(['B_fighter', 'R_odds', 'B_odds', 'R_ev', 'B_ev', 'date', 'location',
       'country', 'title_bout', 'weight_class', 'gender', 'no_of_rounds',
       'B_current_lose_streak', 'B_current_win_streak', 'B_draw',
       'B_avg_SIG_STR_landed', 'B_avg_SIG_STR_pct', 'B_avg_SUB_ATT',
       'B_avg_TD_landed'],
      dtype='object')

In [66]:
## checking the columns which has null values greater than or equal to 7
for i in data.columns:
    if data[i].isnull().sum()>=7:
        print(i)

B_match_weightclass_rank
R_match_weightclass_rank
R_Women's Flyweight_rank
R_Women's Featherweight_rank
R_Women's Strawweight_rank
R_Women's Bantamweight_rank
R_Heavyweight_rank
R_Light Heavyweight_rank
R_Middleweight_rank
R_Welterweight_rank
R_Lightweight_rank
R_Featherweight_rank
R_Bantamweight_rank
R_Flyweight_rank
R_Pound-for-Pound_rank
B_Women's Flyweight_rank
B_Women's Featherweight_rank
B_Women's Strawweight_rank
B_Women's Bantamweight_rank
B_Heavyweight_rank
B_Light Heavyweight_rank
B_Middleweight_rank
B_Welterweight_rank
B_Lightweight_rank
B_Featherweight_rank
B_Bantamweight_rank
B_Flyweight_rank
B_Pound-for-Pound_rank
finish
finish_details
finish_round
finish_round_time
total_fight_time_secs


In [100]:
## deleting all columns which has missing values greter than 7 or equal to 7
for i in data.columns:
    if data[i].isnull().sum()>=7:
        data.drop(columns = [i],inplace = True)
data.shape


(9, 85)

In [67]:
#checking more rest of the columns which has missing values 
data.isna().sum()[40:86]


R_avg_SIG_STR_pct               0
R_avg_SUB_ATT                   0
R_avg_TD_landed                 0
R_avg_TD_pct                    0
R_longest_win_streak            0
R_losses                        0
R_total_rounds_fought           0
R_total_title_bouts             0
R_win_by_Decision_Majority      0
R_win_by_Decision_Split         0
R_win_by_Decision_Unanimous     0
R_win_by_KO/TKO                 0
R_win_by_Submission             0
R_win_by_TKO_Doctor_Stoppage    0
R_wins                          0
R_Stance                        0
R_Height_cms                    0
R_Reach_cms                     0
R_Weight_lbs                    0
R_age                           0
B_age                           0
lose_streak_dif                 0
win_streak_dif                  0
longest_win_streak_dif          0
win_dif                         0
loss_dif                        0
total_round_dif                 0
total_title_bout_dif            0
ko_dif                          0
sub_dif       

In [68]:
# this command is used to delete all those columns which has at least one missing value
data.dropna(axis=1,inplace=True)
data.shape

(9, 82)

In [72]:
# calculating percentage of missing values of all the columns based on its percentage we delete the columns like mostly \
#delete 75% missing value column data considered as irrelavant rest the columns must be cleaned by fillna\
#if they have 1 or 2 missing value
import seaborn as sns
import numpy as np
data = sns.load_dataset('titanic')
data.shape
(data.isnull().sum()/data.shape[0])*100    # showing % of missing values of all the columns like deck has 77% missing values


survived        0.000000
pclass          0.000000
sex             0.000000
age            19.865320
sibsp           0.000000
parch           0.000000
fare            0.000000
embarked        0.224467
class           0.000000
who             0.000000
adult_male      0.000000
deck           77.216611
embark_town     0.224467
alive           0.000000
alone           0.000000
dtype: float64

### deleting nan values using dropna


In [89]:
##Dropping Rows with at least 1 null value.use : axis=0 and how='any'

#Data frame is read and all rows with any Null values are dropped. 

import numpy as np
import pandas as pd
df=pd.DataFrame({"A":[1,2,np.nan,5],"B":[5,np.nan,np.nan,7],"c":[1,2,np.nan,9]})  #np.nan is used to add missing values 
print(df)

# making new data frame with dropped NA values
new_df = df.dropna(axis = 0, how ='any')
  
print(new_df)

     A    B    c
0  1.0  5.0  1.0
1  2.0  NaN  2.0
2  NaN  NaN  NaN
3  5.0  7.0  9.0
     A    B    c
0  1.0  5.0  1.0
3  5.0  7.0  9.0


In [90]:
##Dropping Rows with all null values.  use : axis=0 and how='all'



import numpy as np
import pandas as pd
df=pd.DataFrame({"A":[1,2,np.nan,5],"B":[5,np.nan,np.nan,7],"c":[1,2,np.nan,9]})  #np.nan is used to add missing values 
print(df)

# making new data frame with dropped NA values
new_df = df.dropna(axis = 0, how ='all')
  
print(new_df)

     A    B    c
0  1.0  5.0  1.0
1  2.0  NaN  2.0
2  NaN  NaN  NaN
3  5.0  7.0  9.0
     A    B    c
0  1.0  5.0  1.0
1  2.0  NaN  2.0
3  5.0  7.0  9.0


In [91]:
##Dropping columns with at least 1 null value. use : axis=1 and how='any'

#Data frame is read and all rows with any Null values are dropped. 

import numpy as np
import pandas as pd
df=pd.DataFrame({"A":[1,2,np.nan,5],"B":[5,np.nan,np.nan,np.nan],"c":[1,2,4,9]})  #np.nan is used to add missing values 
print(df)

# making new data frame with dropped NA values
new_df = df.dropna(axis = 1, how ='any')
  
print(new_df)

     A    B  c
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  4
3  5.0  NaN  9
   c
0  1
1  2
2  4
3  9


In [1]:
##Dropping columns with all null value. use : axix =1 and how=all


import numpy as np
import pandas as pd
df=pd.DataFrame({"A":[1,2,np.nan,5],"B":[np.nan,np.nan,np.nan,np.nan],"c":[1,2,4,9]})  #np.nan is used to add missing values 
print(df)

# making new data frame with dropped NA values
new_df = df.dropna(axis = 1, how ='all')
  
print(new_df)

     A   B  c
0  1.0 NaN  1
1  2.0 NaN  2
2  NaN NaN  4
3  5.0 NaN  9
     A  c
0  1.0  1
1  2.0  2
2  NaN  4
3  5.0  9
